// @vercel/microfrontends/dist/experimental/vite.js
// src/config/microfrontends/server/index.ts
import fs6 from "node:fs";
import { dirname as dirname3 } from "node:path";

// src/config/overrides/constants.ts
var OVERRIDES_COOKIE_PREFIX = "vercel-micro-frontends-override";
var OVERRIDES_ENV_COOKIE_PREFIX = `${OVERRIDES_COOKIE_PREFIX}:env:`;

// src/config/overrides/is-override-cookie.ts
function isOverrideCookie(cookie) {
  return Boolean(cookie.name?.startsWith(OVERRIDES_COOKIE_PREFIX));
}

// src/config/overrides/get-override-from-cookie.ts
function getOverrideFromCookie(cookie) {
  if (!isOverrideCookie(cookie) || !cookie.value)
    return;
  return {
    application: cookie.name.replace(OVERRIDES_ENV_COOKIE_PREFIX, ""),
    host: cookie.value
  };
}

// src/config/overrides/parse-overrides.ts
function parseOverrides(cookies) {
  const overridesConfig = { applications: {} };
  cookies.forEach((cookie) => {
    const override = getOverrideFromCookie(cookie);
    if (!override)
      return;
    overridesConfig.applications[override.application] = {
      environment: { host: override.host }
    };
  });
  return overridesConfig;
}

// src/config/errors.ts
var MicrofrontendError = class extends Error {
  constructor(message, opts) {
    super(message, { cause: opts?.cause });
    this.name = "MicrofrontendsError";
    this.source = opts?.source ?? "@vercel/microfrontends";
    this.type = opts?.type ?? "unknown";
    this.subtype = opts?.subtype;
    Error.captureStackTrace(this, MicrofrontendError);
  }
  isKnown() {
    return this.type !== "unknown";
  }
  isUnknown() {
    return !this.isKnown();
  }
  /**
   * Converts an error to a MicrofrontendsError.
   * @param original - The original error to convert.
   * @returns The converted MicrofrontendsError.
   */
  static convert(original, opts) {
    if (opts?.fileName) {
      const err = MicrofrontendError.convertFSError(original, opts.fileName);
      if (err) {
        return err;
      }
    }
    if (original.message.includes(
      "Code generation from strings disallowed for this context"
    )) {
      return new MicrofrontendError(original.message, {
        type: "config",
        subtype: "unsupported_validation_env",
        source: "ajv"
      });
    }
    return new MicrofrontendError(original.message);
  }
  static convertFSError(original, fileName) {
    if (original instanceof Error && "code" in original) {
      if (original.code === "ENOENT") {
        return new MicrofrontendError(`Could not find "${fileName}"`, {
          type: "config",
          subtype: "unable_to_read_file",
          source: "fs"
        });
      }
      if (original.code === "EACCES") {
        return new MicrofrontendError(
          `Permission denied while accessing "${fileName}"`,
          {
            type: "config",
            subtype: "invalid_permissions",
            source: "fs"
          }
        );
      }
    }
    if (original instanceof SyntaxError) {
      return new MicrofrontendError(
        `Failed to parse "${fileName}": Invalid JSON format.`,
        {
          type: "config",
          subtype: "invalid_syntax",
          source: "fs"
        }
      );
    }
    return null;
  }
  /**
   * Handles an unknown error and returns a MicrofrontendsError instance.
   * @param err - The error to handle.
   * @returns A MicrofrontendsError instance.
   */
  static handle(err, opts) {
    if (err instanceof MicrofrontendError) {
      return err;
    }
    if (err instanceof Error) {
      return MicrofrontendError.convert(err, opts);
    }
    if (typeof err === "object" && err !== null) {
      if ("message" in err && typeof err.message === "string") {
        return MicrofrontendError.convert(new Error(err.message), opts);
      }
    }
    return new MicrofrontendError("An unknown error occurred");
  }
};

// src/config/microfrontends-config/utils/get-config-from-env.ts
function getConfigStringFromEnv() {
  const config = process.env.MFE_CONFIG;
  if (!config) {
    throw new MicrofrontendError(`Missing "MFE_CONFIG" in environment.`, {
      type: "config",
      subtype: "not_found_in_env"
    });
  }
  return config;
}

// src/config/microfrontends-config/isomorphic/index.ts
import { parse } from "jsonc-parser";

// src/config/schema/utils/is-main-config.ts
function isMainConfig(c) {
  return !("partOf" in c);
}

// src/config/schema/utils/is-default-app.ts
function isDefaultApp(a) {
  return !("routing" in a);
}

// src/config/microfrontends-config/client/index.ts
import { pathToRegexp } from "path-to-regexp";
var MicrofrontendConfigClient = class {
  constructor(config, opts) {
    this.pathCache = {};
    this.serialized = config;
    if (opts?.removeFlaggedPaths) {
      for (const app of Object.values(config.applications)) {
        if (app.routing) {
          app.routing = app.routing.filter((match) => !match.flag);
        }
      }
    }
    this.applications = config.applications;
  }
  /**
   * Create a new `MicrofrontendConfigClient` from a JSON string.
   * Config must be passed in to remain framework agnostic
   */
  static fromEnv(config, opts) {
    if (!config) {
      throw new Error("No microfrontends configuration found");
    }
    return new MicrofrontendConfigClient(
      JSON.parse(config),
      opts
    );
  }
  isEqual(other) {
    return JSON.stringify(this.applications) === JSON.stringify(other.applications);
  }
  getApplicationNameForPath(path6) {
    if (!path6.startsWith("/")) {
      throw new Error(`Path must start with a /`);
    }
    if (this.pathCache[path6]) {
      return this.pathCache[path6];
    }
    const pathname = new URL(path6, "https://example.com").pathname;
    for (const [name, application] of Object.entries(this.applications)) {
      if (application.routing) {
        for (const group of application.routing) {
          for (const childPath of group.paths) {
            const regexp = pathToRegexp(childPath);
            if (regexp.test(pathname)) {
              this.pathCache[path6] = name;
              return name;
            }
          }
        }
      }
    }
    const defaultApplication = Object.entries(this.applications).find(
      ([, application]) => application.default
    );
    if (!defaultApplication) {
      return null;
    }
    this.pathCache[path6] = defaultApplication[0];
    return defaultApplication[0];
  }
  serialize() {
    return this.serialized;
  }
};

// src/config/microfrontends-config/isomorphic/validation.ts
import { pathToRegexp as pathToRegexp2, parse as parsePathRegexp } from "path-to-regexp";
var validateConfigPaths = (applicationConfigsById) => {
  if (!applicationConfigsById) {
    return;
  }
  const pathsByApplicationId = /* @__PURE__ */ new Map();
  const errors = [];
  for (const [id, app] of Object.entries(applicationConfigsById)) {
    if (isDefaultApp(app)) {
      continue;
    }
    const childApp = app;
    for (const pathMatch of childApp.routing) {
      for (const path6 of pathMatch.paths) {
        const maybeError = validatePathExpression(path6);
        if (maybeError) {
          errors.push(maybeError);
        } else {
          const existing = pathsByApplicationId.get(path6);
          if (existing) {
            existing.applications.push(id);
          } else {
            pathsByApplicationId.set(path6, {
              applications: [id],
              matcher: pathToRegexp2(path6),
              applicationId: id
            });
          }
        }
      }
    }
  }
  const entries = Array.from(pathsByApplicationId.entries());
  for (const [path6, { applications: ids, matcher, applicationId }] of entries) {
    if (ids.length > 1) {
      errors.push(
        `Duplicate path "${path6}" for applications "${ids.join(", ")}"`
      );
    }
    for (const [
      matchPath,
      { applications: matchIds, applicationId: matchApplicationId }
    ] of entries) {
      if (path6 === matchPath) {
        continue;
      }
      if (applicationId === matchApplicationId) {
        continue;
      }
      if (matcher.test(matchPath)) {
        const source = `"${path6}" of application${ids.length > 0 ? "s" : ""} ${ids.join(", ")}`;
        const destination = `"${matchPath}" of application${matchIds.length > 0 ? "s" : ""} ${matchIds.join(", ")}`;
        errors.push(
          `Overlapping path detected between ${source} and ${destination}`
        );
      }
    }
  }
  if (errors.length) {
    throw new MicrofrontendError(`Invalid paths: ${errors.join(", ")}`, {
      type: "config",
      subtype: "conflicting_paths"
    });
  }
};
var PATH_DEFAULT_PATTERN = "[^\\/#\\?]+?";
function validatePathExpression(path6) {
  try {
    const tokens = parsePathRegexp(path6);
    if (/(?<!\\)\{/.test(path6)) {
      return `Optional paths are not supported: ${path6}`;
    }
    if (/(?<!\\|\()\?/.test(path6)) {
      return `Optional paths are not supported: ${path6}`;
    }
    if (/\/[^/]*(?<!\\):[^/]*(?<!\\):[^/]*/.test(path6)) {
      return `Only one wildcard is allowed per path segment: ${path6}`;
    }
    for (let i = 0; i < tokens.length; i++) {
      const token = tokens[i];
      if (token === void 0) {
        return `token ${i} in ${path6} is undefined, this shouldn't happen`;
      }
      if (typeof token !== "string") {
        if (!token.name) {
          return `Only named wildcards are allowed: ${path6} (hint: add ":path" to the wildcard)`;
        }
        if (token.pattern !== PATH_DEFAULT_PATTERN && // Allows (a|b|c) and ((?!a|b|c).*) regex
        // Only limited regex is supported for now, due to performance considerations
        !/^(?<allowed>[\w]+(?:\|[^|()]+)+)$|^\(\?!(?<disallowed>[\w]+(?:\|[^|()]+)+)\)\.\*$/.test(
          token.pattern
        )) {
          return `Path ${path6} cannot use unsupported regular expression wildcard`;
        }
        if (token.modifier && i !== tokens.length - 1) {
          return `Modifier ${token.modifier} is not allowed on wildcard :${token.name} in ${path6}. Modifiers are only allowed in the last path component`;
        }
      }
    }
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    return `Path ${path6} could not be parsed into regexp: ${message}`;
  }
  return void 0;
}
var validateAppPaths = (name, app) => {
  for (const group of app.routing) {
    for (const p of group.paths) {
      if (p === "/") {
        continue;
      }
      if (p.endsWith("/")) {
        throw new MicrofrontendError(
          `Invalid path for application "${name}". ${p} must not end with a slash.`,
          { type: "application", subtype: "invalid_path" }
        );
      }
      if (!p.startsWith("/")) {
        throw new MicrofrontendError(
          `Invalid path for application "${name}". ${p} must start with a slash.`,
          { type: "application", subtype: "invalid_path" }
        );
      }
    }
  }
};
var validateConfigDefaultApplication = (applicationConfigsById) => {
  if (!applicationConfigsById) {
    return;
  }
  const applicationsWithRouting = Object.entries(applicationConfigsById).filter(
    ([, app]) => !isDefaultApp(app)
  );
  const applicationsWithRoutingNames = applicationsWithRouting.map(
    ([key]) => key
  );
  const numApplications = Object.keys(applicationConfigsById).length;
  const numApplicationsWithRouting = applicationsWithRoutingNames.length;
  const numApplicationsWithoutRouting = numApplications - numApplicationsWithRouting;
  if (numApplicationsWithoutRouting === 0) {
    throw new MicrofrontendError(
      "No default application found. At least one application needs to be the default by omitting routing.",
      { type: "config", subtype: "no_default_application" }
    );
  }
  if (numApplicationsWithoutRouting > 1) {
    throw new MicrofrontendError(
      `Only one application can omit "routing". Found ${applicationsWithRoutingNames.length - Object.keys(applicationConfigsById).length > 1}.`,
      { type: "config", subtype: "multiple_default_applications" }
    );
  }
};
var validateDeprecatedFields = (config) => {
  const errors = [];
  if (config.options?.vercel) {
    errors.push(
      `Configuration cannot contain deprecated field 'options.vercel'. Use 'options.disableOverrides' instead.`
    );
  }
  if (config.options?.localProxy) {
    errors.push(
      `Configuration cannot contain deprecated field 'options.localProxy'. Use 'options.localProxyPort' instead.`
    );
  }
  for (const [applicationId, application] of Object.entries(
    config.applications
  )) {
    if (application.vercel) {
      errors.push(
        `Application '${applicationId}' cannot contain deprecated field 'vercel'. Use 'projectId' instead.`
      );
    }
    if (application.production) {
      errors.push(
        `Application '${applicationId}' cannot contain deprecated field 'production'. Use 'development.fallback' instead.`
      );
    }
    if (application.development?.local) {
      errors.push(
        `Application '${applicationId}' cannot contain deprecated field 'development.local'. Use 'developement.localPort' instead.`
      );
    }
  }
  if (errors.length) {
    throw new MicrofrontendError(
      `Microfrontends configuration file errors:
- ${errors.join("\n- ")}`,
      {
        type: "config",
        subtype: "depcrecated_field"
      }
    );
  }
};

// src/config/microfrontends-config/isomorphic/utils/generate-asset-prefix.ts
var PREFIX = "vc-ap";
function generateAssetPrefixFromName({
  name
}) {
  if (!name) {
    throw new Error("Name is required to generate an asset prefix");
  }
  return `${PREFIX}-${name}`;
}

// src/config/microfrontends-config/isomorphic/utils/generate-port.ts
function generatePortFromName({
  name,
  minPort = 3e3,
  maxPort = 8e3
}) {
  if (!name) {
    throw new Error("Name is required to generate a port");
  }
  let hash = 0;
  for (let i = 0; i < name.length; i++) {
    hash = (hash << 5) - hash + name.charCodeAt(i);
    hash |= 0;
  }
  hash = Math.abs(hash);
  const range = maxPort - minPort;
  const port = minPort + hash % range;
  return port;
}

// src/config/microfrontends-config/isomorphic/host.ts
var Host = class {
  constructor(hostConfig, options) {
    if (typeof hostConfig === "string") {
      ({
        protocol: this.protocol,
        host: this.host,
        port: this.port
      } = Host.parseUrl(hostConfig));
    } else {
      const { protocol = "https", host, port } = hostConfig;
      this.protocol = protocol;
      this.host = host;
      this.port = port;
    }
    this.local = options?.isLocal;
  }
  static parseUrl(url) {
    let hostToParse = url;
    if (!/^https?:\/\//.exec(hostToParse)) {
      hostToParse = `https://${hostToParse}`;
    }
    const parsed = new URL(hostToParse);
    if (!parsed.hostname) {
      throw new Error(Host.getMicrofrontendsError(url, "requires a host"));
    }
    if (parsed.hash) {
      throw new Error(
        Host.getMicrofrontendsError(url, "cannot have a fragment")
      );
    }
    if (parsed.username || parsed.password) {
      throw new Error(
        Host.getMicrofrontendsError(
          url,
          "cannot have authentication credentials (username and/or password)"
        )
      );
    }
    if (parsed.pathname !== "/") {
      throw new Error(Host.getMicrofrontendsError(url, "cannot have a path"));
    }
    if (parsed.search) {
      throw new Error(
        Host.getMicrofrontendsError(url, "cannot have query parameters")
      );
    }
    const protocol = parsed.protocol.slice(0, -1);
    return {
      protocol,
      host: parsed.hostname,
      port: parsed.port ? Number.parseInt(parsed.port) : void 0
    };
  }
  static getMicrofrontendsError(url, message) {
    return `Microfrontends configuration error: the URL ${url} in your microfrontends.json ${message}.`;
  }
  isLocal() {
    return this.local || this.host === "localhost" || this.host === "127.0.0.1";
  }
  toString() {
    const url = this.toUrl();
    return url.toString().replace(/\/$/, "");
  }
  toUrl() {
    const url = `${this.protocol}://${this.host}${this.port ? `:${this.port}` : ""}`;
    return new URL(url);
  }
};
var LocalHost = class extends Host {
  constructor({
    appName,
    localPort,
    ...hostConfig
  }) {
    const host = hostConfig.host ?? "localhost";
    const port = localPort ?? hostConfig.port ?? generatePortFromName({ name: appName });
    const protocol = hostConfig.protocol ?? "http";
    super({ protocol, host, port });
  }
};

// src/config/microfrontends-config/isomorphic/application.ts
var Application = class {
  constructor(name, {
    app,
    overrides,
    isDefault
  }) {
    this.name = name;
    this.development = {
      local: new LocalHost({
        appName: name,
        localPort: app.development?.localPort,
        ...app.development?.local
      }),
      fallback: app.development?.fallback ? new Host(app.development.fallback) : void 0
    };
    if (app.development?.fallback) {
      this.fallback = new Host(app.development.fallback);
    } else if (app.production) {
      this.fallback = new Host(app.production);
    }
    this.projectId = app.projectId ?? app.vercel?.projectId;
    this.overrides = overrides?.environment ? {
      environment: new Host(overrides.environment)
    } : void 0;
    this.default = isDefault ?? false;
    this.serialized = app;
  }
  isDefault() {
    return this.default;
  }
  getAssetPrefix() {
    return generateAssetPrefixFromName({ name: this.name });
  }
  serialize() {
    return this.serialized;
  }
};
var DefaultApplication = class extends Application {
  constructor(name, {
    app,
    overrides
  }) {
    super(name, {
      app,
      overrides,
      isDefault: true
    });
    this.default = true;
    const fallbackHost = app.development?.fallback ?? app.production;
    if (fallbackHost === void 0) {
      throw new Error(
        "`app.production` or `app.development.fallback` must be set in the default application in microfrontends.json."
      );
    }
    this.fallback = new Host(fallbackHost);
    if (app.production) {
      this.production = new Host(app.production);
    }
  }
  getAssetPrefix() {
    return "";
  }
};
var ChildApplication = class extends Application {
  constructor(name, {
    app,
    overrides
  }) {
    ChildApplication.validate(name, app);
    super(name, {
      app,
      overrides,
      isDefault: false
    });
    this.default = false;
    this.routing = app.routing;
  }
  static validate(name, app) {
    validateAppPaths(name, app);
  }
};

// src/config/microfrontends-config/isomorphic/constants.ts
var DEFAULT_LOCAL_PROXY_PORT = 3024;

// src/config/microfrontends-config/isomorphic/index.ts
var MicrofrontendConfigIsomorphic = class {
  constructor({
    config,
    overrides,
    meta,
    opts
  }) {
    this.childApplications = {};
    MicrofrontendConfigIsomorphic.validate(config, opts);
    const disableOverrides = config.options?.disableOverrides ?? config.options?.vercel?.disableOverrides ?? false;
    this.overrides = overrides && !disableOverrides ? overrides : void 0;
    this.isMainConfig = isMainConfig(config);
    if (isMainConfig(config)) {
      for (const [appId, appConfig] of Object.entries(config.applications)) {
        const appOverrides = !disableOverrides ? this.overrides?.applications[appId] : void 0;
        if (isDefaultApp(appConfig)) {
          this.defaultApplication = new DefaultApplication(appId, {
            app: appConfig,
            overrides: appOverrides
          });
        } else {
          this.childApplications[appId] = new ChildApplication(appId, {
            app: appConfig,
            overrides: appOverrides
          });
        }
      }
    } else {
      this.partOf = config.partOf;
      const appOverrides = !disableOverrides ? this.overrides?.applications[meta.fromApp] : void 0;
      this.childApplications[meta.fromApp] = new ChildApplication(
        meta.fromApp,
        {
          // we don't know routing because we're not in the main config
          app: { routing: [] },
          overrides: appOverrides
        }
      );
    }
    if (isMainConfig(config) && !this.defaultApplication) {
      throw new MicrofrontendError(
        "Could not find default application in microfrontends configuration",
        {
          type: "application",
          subtype: "not_found"
        }
      );
    }
    this.config = config;
    this.options = config.options;
    this.serialized = {
      config,
      overrides,
      meta
    };
  }
  static validate(config, opts) {
    const skipValidation = opts?.skipValidation ?? [];
    const c = typeof config === "string" ? parse(config) : config;
    if (isMainConfig(c)) {
      validateConfigPaths(c.applications);
      validateConfigDefaultApplication(c.applications);
      if (!skipValidation.includes("deprecatedFields")) {
        validateDeprecatedFields(c);
      }
    }
    return c;
  }
  static fromEnv({
    meta,
    cookies
  }) {
    return new MicrofrontendConfigIsomorphic({
      config: parse(getConfigStringFromEnv()),
      overrides: parseOverrides(cookies ?? []),
      meta
    });
  }
  isOverridesDisabled() {
    return this.options?.vercel?.disableOverrides ?? false;
  }
  getConfig() {
    return this.config;
  }
  getApplicationsByType() {
    return {
      defaultApplication: this.defaultApplication,
      applications: Object.values(this.childApplications)
    };
  }
  getChildApplications() {
    return Object.values(this.childApplications);
  }
  getAllApplications() {
    return [
      this.defaultApplication,
      ...Object.values(this.childApplications)
    ].filter(Boolean);
  }
  getApplication(name) {
    if (this.defaultApplication?.name === name) {
      return this.defaultApplication;
    }
    const app = this.childApplications[name];
    if (!app) {
      throw new MicrofrontendError(
        `Could not find microfrontends configuration for application "${name}"`,
        {
          type: "application",
          subtype: "not_found"
        }
      );
    }
    return app;
  }
  getApplicationByProjectId(projectId) {
    if (this.defaultApplication?.projectId === projectId) {
      return this.defaultApplication;
    }
    return Object.values(this.childApplications).find(
      (app) => app.projectId === projectId
    );
  }
  /**
   * Returns the default application. This can throw if the default application
   * is undefined ( )
   */
  getDefaultApplication() {
    if (!this.defaultApplication) {
      throw new MicrofrontendError(
        "Could not find default application in microfrontends configuration",
        {
          type: "application",
          subtype: "not_found"
        }
      );
    }
    return this.defaultApplication;
  }
  /**
   * Returns the configured port for the local proxy
   */
  getLocalProxyPort() {
    return this.config.options?.localProxyPort ?? this.config.options?.localProxy?.port ?? DEFAULT_LOCAL_PROXY_PORT;
  }
  /**
   * Serializes the class back to the Schema type.
   *
   * NOTE: This is used when writing the config to disk and must always match the input Schema
   */
  toSchemaJson() {
    return this.serialized.config;
  }
  toClientConfig() {
    const applications = Object.fromEntries(
      Object.entries(this.childApplications).map(([name, application]) => [
        name,
        {
          default: false,
          routing: application.routing
        }
      ])
    );
    if (this.defaultApplication) {
      applications[this.defaultApplication.name] = {
        default: true
      };
    }
    return new MicrofrontendConfigClient({
      applications
    });
  }
  serialize() {
    return this.serialized;
  }
};

// src/config/microfrontends-config/isomorphic/child.ts
var MicrofrontendChildConfig = class extends MicrofrontendConfigIsomorphic {
  constructor({
    config,
    overrides,
    meta
  }) {
    super({ config, overrides, meta });
    this.isMainConfig = false;
    this.partOf = config.partOf;
  }
};

// src/config/microfrontends-config/isomorphic/main.ts
var MicrofrontendMainConfig = class extends MicrofrontendConfigIsomorphic {
  constructor({
    config,
    overrides,
    meta
  }) {
    super({ config, overrides, meta });
    this.isMainConfig = true;
    const disableOverrides = config.options?.disableOverrides ?? config.options?.vercel?.disableOverrides ?? false;
    let defaultApplication;
    for (const [appId, appConfig] of Object.entries(config.applications)) {
      const appOverrides = !disableOverrides ? this.overrides?.applications[appId] : void 0;
      if (isDefaultApp(appConfig)) {
        defaultApplication = new DefaultApplication(appId, {
          app: appConfig,
          overrides: appOverrides
        });
      } else {
        this.childApplications[appId] = new ChildApplication(appId, {
          app: appConfig,
          overrides: appOverrides
        });
      }
    }
    if (!defaultApplication) {
      throw new MicrofrontendError(
        "Could not find default application in microfrontends configuration",
        {
          type: "application",
          subtype: "not_found"
        }
      );
    }
    this.defaultApplication = defaultApplication;
  }
};

// src/config/microfrontends/isomorphic/index.ts
var Microfrontends = class {
  constructor({
    config,
    overrides,
    meta
  }) {
    if (isMainConfig(config)) {
      this.config = new MicrofrontendMainConfig({ config, overrides, meta });
    } else {
      this.config = new MicrofrontendChildConfig({ config, overrides, meta });
    }
  }
  isChildConfig() {
    return this.config instanceof MicrofrontendChildConfig;
  }
  static fromEnv({
    cookies,
    meta
  }) {
    const config = MicrofrontendConfigIsomorphic.fromEnv({
      cookies,
      meta
    });
    return new Microfrontends(config.serialize());
  }
};

// src/config/microfrontends/utils/find-repository-root.ts
import fs from "node:fs";
import path from "node:path";
var GIT_DIRECTORY = ".git";
function findRepositoryRoot(startDir) {
  if (process.env.NX_WORKSPACE_ROOT) {
    return process.env.NX_WORKSPACE_ROOT;
  }
  let currentDir = startDir || process.cwd();
  while (currentDir !== path.parse(currentDir).root) {
    const gitPath = path.join(currentDir, GIT_DIRECTORY);
    if (fs.existsSync(gitPath) && fs.statSync(gitPath).isDirectory()) {
      return currentDir;
    }
    currentDir = path.dirname(currentDir);
  }
  throw new Error(
    "Repository root not found. Specify the root of the repository with the `repository.root` option."
  );
}

// src/config/microfrontends/utils/find-package-path.ts
import { dirname } from "node:path";
import { readFileSync } from "node:fs";
import fg from "fast-glob";
var configCache = {};
function findPackagePathWithGlob({
  repositoryRoot,
  name
}) {
  try {
    const packageJsonPaths = fg.globSync("**/package.json", {
      cwd: repositoryRoot,
      absolute: true,
      onlyFiles: true,
      followSymbolicLinks: false,
      ignore: ["**/node_modules/**", "**/.git/**"]
    });
    const matchingPaths = [];
    for (const packageJsonPath2 of packageJsonPaths) {
      const packageJsonContent = readFileSync(packageJsonPath2, "utf-8");
      const packageJson = JSON.parse(packageJsonContent);
      if (packageJson.name === name) {
        matchingPaths.push(packageJsonPath2);
      }
    }
    if (matchingPaths.length > 1) {
      throw new Error(
        `Found multiple packages with the name "${name}" in the repository: ${matchingPaths.join(", ")}`
      );
    }
    if (matchingPaths.length === 0) {
      throw new Error(
        `Could not find package with the name "${name}" in the repository`
      );
    }
    const [packageJsonPath] = matchingPaths;
    return dirname(packageJsonPath);
  } catch (error) {
    return null;
  }
}
function findPackagePath(opts) {
  const cacheKey = `${opts.repositoryRoot}-${opts.name}`;
  if (configCache[cacheKey]) {
    return configCache[cacheKey];
  }
  const result = findPackagePathWithGlob(opts);
  if (!result) {
    throw new Error(
      `Could not find package with the name "${opts.name}" in the repository`
    );
  }
  configCache[cacheKey] = result;
  return result;
}

// src/config/microfrontends/utils/find-default-package.ts
import { dirname as dirname2 } from "node:path";
import { readFileSync as readFileSync2 } from "node:fs";
import { parse as parse2 } from "jsonc-parser";
import fg2 from "fast-glob";

// src/config/constants.ts
var CONFIGURATION_FILENAMES = [
  "microfrontends.jsonc",
  "microfrontends.json"
];

// src/config/microfrontends/utils/find-default-package.ts
var configCache2 = {};
function findDefaultMicrofrontendsPackages({
  repositoryRoot,
  applicationName
}) {
  try {
    const microfrontendsJsonPaths = fg2.globSync(
      `**/{${CONFIGURATION_FILENAMES.join(",")}}`,
      {
        cwd: repositoryRoot,
        absolute: true,
        onlyFiles: true,
        followSymbolicLinks: false,
        ignore: ["**/node_modules/**", "**/.git/**"]
      }
    );
    const matchingPaths = [];
    for (const microfrontendsJsonPath of microfrontendsJsonPaths) {
      try {
        const microfrontendsJsonContent = readFileSync2(
          microfrontendsJsonPath,
          "utf-8"
        );
        const microfrontendsJson = parse2(microfrontendsJsonContent);
        if (isMainConfig(microfrontendsJson) && microfrontendsJson.applications[applicationName]) {
          matchingPaths.push(microfrontendsJsonPath);
        }
      } catch (error) {
      }
    }
    if (matchingPaths.length > 1) {
      throw new Error(
        `Found multiple default applications referencing "${applicationName}" in the repository, this is not yet supported.
${matchingPaths.join("\n  \u2022 ")}`
      );
    }
    if (matchingPaths.length === 0) {
      throw new Error(
        `Could not find default application with "applications.${applicationName}"`
      );
    }
    const [packageJsonPath] = matchingPaths;
    return dirname2(packageJsonPath);
  } catch (error) {
    return null;
  }
}
function findDefaultMicrofrontendsPackage(opts) {
  const cacheKey = `${opts.repositoryRoot}-${opts.applicationName}`;
  if (configCache2[cacheKey]) {
    return configCache2[cacheKey];
  }
  const result = findDefaultMicrofrontendsPackages(opts);
  if (!result) {
    throw new Error(
      "Error trying to resolve the main microfrontends configuration"
    );
  }
  configCache2[cacheKey] = result;
  return result;
}

// src/config/microfrontends/utils/is-monorepo.ts
import fs2 from "node:fs";
import path2 from "node:path";
function isMonorepo({
  repositoryRoot
}) {
  try {
    if (fs2.existsSync(path2.join(repositoryRoot, "pnpm-workspace.yaml"))) {
      return true;
    }
    if (fs2.existsSync(path2.join(repositoryRoot, "vlt-workspaces.json"))) {
      return true;
    }
    if (process.env.NX_WORKSPACE_ROOT === path2.resolve(repositoryRoot)) {
      return true;
    }
    const packageJsonPath = path2.join(repositoryRoot, "package.json");
    if (!fs2.existsSync(packageJsonPath)) {
      return false;
    }
    const packageJson = JSON.parse(
      fs2.readFileSync(packageJsonPath, "utf-8")
    );
    return packageJson.workspaces !== void 0;
  } catch (error) {
    console.error("Error determining if repository is a monorepo", error);
    return false;
  }
}

// src/config/microfrontends/utils/find-package-root.ts
import fs3 from "node:fs";
import path3 from "node:path";
var PACKAGE_JSON = "package.json";
function findPackageRoot(startDir) {
  let currentDir = startDir || process.cwd();
  while (currentDir !== path3.parse(currentDir).root) {
    const pkgJsonPath = path3.join(currentDir, PACKAGE_JSON);
    if (fs3.existsSync(pkgJsonPath)) {
      return currentDir;
    }
    currentDir = path3.dirname(currentDir);
  }
  throw new Error(
    "Package root not found. Specify the root of the package with the `package.root` option."
  );
}

// src/config/microfrontends/utils/find-config.ts
import fs4 from "node:fs";
import { join } from "node:path";
function findConfig({ dir }) {
  for (const filename of CONFIGURATION_FILENAMES) {
    const maybeConfig = join(dir, filename);
    if (fs4.existsSync(maybeConfig)) {
      return maybeConfig;
    }
  }
  return null;
}

// src/config/microfrontends/utils/get-application-context.ts
import fs5 from "node:fs";
import path4 from "node:path";
function getApplicationContext(opts) {
  if (opts?.appName) {
    return { name: opts.appName };
  }
  if (process.env.NX_TASK_TARGET_PROJECT) {
    return { name: process.env.NX_TASK_TARGET_PROJECT };
  }
  try {
    const packageJsonString = fs5.readFileSync(
      path4.join(opts?.packageRoot || ".", "package.json"),
      "utf-8"
    );
    const packageJson = JSON.parse(packageJsonString);
    if (!packageJson.name) {
      throw new MicrofrontendError(
        `package.json file missing required field "name"`,
        {
          type: "packageJson",
          subtype: "missing_field_name",
          source: "@vercel/microfrontends/next"
        }
      );
    }
    return { name: packageJson.name };
  } catch (err) {
    throw MicrofrontendError.handle(err, {
      fileName: "package.json"
    });
  }
}

// src/config/microfrontends/server/utils/get-output-file-path.ts
import path5 from "node:path";

// src/config/microfrontends/server/constants.ts
var MFE_CONFIG_DEFAULT_FILE_PATH = "microfrontends";
var MFE_CONFIG_DEFAULT_FILE_NAME = "microfrontends.json";

// src/config/microfrontends/server/utils/get-output-file-path.ts
function getOutputFilePath() {
  return path5.join(MFE_CONFIG_DEFAULT_FILE_PATH, MFE_CONFIG_DEFAULT_FILE_NAME);
}

// src/config/microfrontends/server/validation.ts
import { parse as parse3 } from "jsonc-parser";
import { Ajv } from "ajv";

// schema/schema.json
var schema_default = {
  $schema: "http://json-schema.org/draft-07/schema#",
  $ref: "#/definitions/Config",
  definitions: {
    Config: {
      anyOf: [
        {
          $ref: "#/definitions/MainConfig"
        },
        {
          $ref: "#/definitions/ChildConfig"
        }
      ]
    },
    MainConfig: {
      type: "object",
      properties: {
        $schema: {
          type: "string"
        },
        version: {
          type: "string",
          const: "1"
        },
        options: {
          $ref: "#/definitions/Options"
        },
        applications: {
          $ref: "#/definitions/ApplicationRouting",
          description: "Mapping of application names to the routes that they host. Only needs to be defined in the application that owns the primary microfrontend domain"
        }
      },
      required: ["applications"],
      additionalProperties: false
    },
    Options: {
      type: "object",
      properties: {
        vercel: {
          $ref: "#/definitions/VercelOptions",
          description: "Microfrontends wide options for Vercel.",
          deprecated: "This is being replaced by the `disableOverrides` field below."
        },
        disableOverrides: {
          type: "boolean",
          description: "If you want to disable the overrides for the site. For example, if you are managing rewrites between applications externally, you may wish to disable the overrides on the toolbar as they will have no effect."
        },
        localProxy: {
          $ref: "#/definitions/LocalProxyOptions",
          description: "Options for local proxy.",
          deprecated: "This is being replaced by the `localProxyPort` field below."
        },
        localProxyPort: {
          type: "number",
          description: "The port number used by the local proxy server.\n\nThe default is `3024`."
        }
      },
      additionalProperties: false
    },
    VercelOptions: {
      type: "object",
      properties: {
        disableOverrides: {
          type: "boolean",
          description: "If you want to disable the overrides for the site. For example, if you are managing rewrites between applications externally, you may wish to disable the overrides on the toolbar as they will have no effect."
        }
      },
      additionalProperties: false
    },
    LocalProxyOptions: {
      type: "object",
      properties: {
        port: {
          type: "number",
          description: "The port number used by the local proxy server.\n\nThe default is `3024`."
        }
      },
      additionalProperties: false
    },
    ApplicationRouting: {
      type: "object",
      additionalProperties: {
        $ref: "#/definitions/Application"
      },
      propertyNames: {
        description: "The unique identifier for a Microfrontend Application. Must match the `name` field of the application's `package.json`."
      }
    },
    Application: {
      anyOf: [
        {
          $ref: "#/definitions/DefaultApplication"
        },
        {
          $ref: "#/definitions/ChildApplication"
        }
      ]
    },
    DefaultApplication: {
      type: "object",
      properties: {
        vercel: {
          $ref: "#/definitions/Vercel",
          deprecated: "This is being replaced by the `projectId` field below."
        },
        projectId: {
          type: "string",
          description: "Vercel project ID"
        },
        production: {
          $ref: "#/definitions/HostConfig",
          deprecated: "This is a duplicate of the `development.fallback` field and this will be removed soon."
        },
        development: {
          $ref: "#/definitions/Development"
        }
      },
      additionalProperties: false
    },
    Vercel: {
      type: "object",
      properties: {
        projectId: {
          type: "string",
          description: "Vercel project ID"
        }
      },
      required: ["projectId"],
      additionalProperties: false
    },
    HostConfig: {
      type: "object",
      properties: {
        protocol: {
          type: "string",
          enum: ["http", "https"],
          description: "The protocol to be used for the connection.\n- `http`: Hypertext Transfer Protocol (HTTP).\n- `https`: Secure Hypertext Transfer Protocol (HTTPS).\n\n*"
        },
        host: {
          type: "string",
          description: "The hostname or IP address of the server. This can be a domain name (e.g., `example.com`) or an IP address (e.g., `192.168.1.1`)."
        },
        port: {
          type: "number",
          description: "The port number to be used for the connection. Common values include `80` for HTTP and `443` for HTTPS."
        }
      },
      required: ["host"],
      additionalProperties: false
    },
    Development: {
      type: "object",
      properties: {
        local: {
          $ref: "#/definitions/LocalHostConfig",
          deprecated: "This is being replaced by the `localPort` field below."
        },
        localPort: {
          type: "number",
          description: "The local port number that this application runs on when it is running locally. Common values include `80` for HTTP and `443` for HTTPS."
        },
        fallback: {
          anyOf: [
            {
              $ref: "#/definitions/HostConfig"
            },
            {
              type: "string"
            }
          ],
          description: "Fallback for local development, could be a host config that points to any environment. If this is not provided, or the application is not running - requests to the application in local development will error.\n\nIf passing a string, include the protocol (optional), host (required) and port (optional). For example: `https://this.ismyhost:8080`. If omitted, the protocol defaults to HTTPS. If omitted, the port defaults to `80` for HTTP and `443` for HTTPS."
        },
        task: {
          type: "string",
          description: "Optional task to run when starting the development server. Should reference a script in the package.json of the application."
        }
      },
      additionalProperties: false
    },
    LocalHostConfig: {
      type: "object",
      additionalProperties: false,
      properties: {
        host: {
          type: "string",
          description: "The hostname or IP address of the server. This can be a domain name (e.g., `example.com`) or an IP address (e.g., `192.168.1.1`)."
        },
        protocol: {
          type: "string",
          enum: ["http", "https"],
          description: "The protocol to be used for the connection.\n- `http`: Hypertext Transfer Protocol (HTTP).\n- `https`: Secure Hypertext Transfer Protocol (HTTPS).\n\n*"
        },
        port: {
          type: "number",
          description: "The port number to be used for the connection. Common values include `80` for HTTP and `443` for HTTPS."
        }
      }
    },
    ChildApplication: {
      type: "object",
      properties: {
        vercel: {
          $ref: "#/definitions/Vercel",
          deprecated: "This is being replaced by the `projectId` field below."
        },
        projectId: {
          type: "string",
          description: "Vercel project ID"
        },
        production: {
          $ref: "#/definitions/HostConfig",
          deprecated: "This is a duplicate of the `development.fallback` field and this will be removed soon."
        },
        development: {
          $ref: "#/definitions/Development"
        },
        routing: {
          $ref: "#/definitions/Routing",
          description: "Groups of path expressions that are routed to this application."
        }
      },
      required: ["routing"],
      additionalProperties: false
    },
    Routing: {
      type: "array",
      items: {
        $ref: "#/definitions/PathGroup"
      }
    },
    PathGroup: {
      type: "object",
      properties: {
        group: {
          type: "string",
          description: "Optional group name for the paths"
        },
        flag: {
          type: "string",
          description: "flag name that can be used to enable/disable all paths in the group"
        },
        paths: {
          type: "array",
          items: {
            type: "string"
          }
        }
      },
      required: ["paths"],
      additionalProperties: false
    },
    ChildConfig: {
      type: "object",
      properties: {
        $schema: {
          type: "string"
        },
        version: {
          type: "string",
          const: "1"
        },
        options: {
          $ref: "#/definitions/Options"
        },
        partOf: {
          type: "string",
          description: "Applications that only serve a subset of the microfrontend routes only need to reference the name of the primary application that owns the full microfrontends configuration."
        }
      },
      required: ["partOf"],
      additionalProperties: false
    }
  }
};

// src/config/schema/utils/load.ts
var SCHEMA = schema_default;

// src/config/microfrontends/server/validation.ts
function filterAjvErrors(errors) {
  if (!errors) {
    return [];
  }
  return errors.filter((error) => {
    return error.keyword === "additionalProperties" || error.keyword === "required";
  });
}
function validateSchema(configString) {
  const parsedConfig = parse3(configString);
  const ajv = new Ajv();
  const validate = ajv.compile(SCHEMA);
  const isValid = validate(parsedConfig);
  if (!isValid) {
    throw new MicrofrontendError(
      `Invalid microfrontends config:
 - ${ajv.errorsText(filterAjvErrors(validate.errors), { separator: "\n - " })}

See https://openapi.vercel.sh/microfrontends.json for the schema.`,
      { type: "config", subtype: "does_not_match_schema" }
    );
  }
  return parsedConfig;
}

// src/config/microfrontends/server/index.ts
var MicrofrontendsServer = class extends Microfrontends {
  /**
   * Writes the configuration to a file.
   */
  writeConfig(opts = {
    pretty: true
  }) {
    const outputPath = getOutputFilePath();
    fs6.mkdirSync(dirname3(outputPath), { recursive: true });
    fs6.writeFileSync(
      outputPath,
      JSON.stringify(
        this.config.toSchemaJson(),
        null,
        opts.pretty ?? true ? 2 : void 0
      )
    );
  }
  // --------- Static Methods ---------
  /**
   * Generates a MicrofrontendsServer instance from an unknown object.
   */
  static fromUnknown({
    config,
    cookies,
    meta
  }) {
    const overrides = cookies ? parseOverrides(cookies) : void 0;
    if (typeof config === "string") {
      return new MicrofrontendsServer({
        config: MicrofrontendsServer.validate(config),
        overrides,
        meta
      });
    }
    if (typeof config === "object") {
      return new MicrofrontendsServer({
        config,
        overrides,
        meta
      });
    }
    throw new MicrofrontendError(
      "Invalid config: must be a string or an object",
      { type: "config", subtype: "does_not_match_schema" }
    );
  }
  /**
   * Generates a MicrofrontendsServer instance from the environment.
   * Uses additional validation that is only available when in a node runtime
   */
  static fromEnv({
    cookies,
    meta
  }) {
    return new MicrofrontendsServer({
      config: MicrofrontendsServer.validate(getConfigStringFromEnv()),
      overrides: parseOverrides(cookies),
      meta
    });
  }
  /**
   * Validates the configuration against the JSON schema
   */
  static validate(config) {
    if (typeof config === "string") {
      const c = validateSchema(config);
      return c;
    }
    return config;
  }
  /**
   * Looks up the configuration by inferring the package root and looking for a microfrontends config file. If a file is not found,
   * it will look for a package in the repository with a microfrontends file that contains the current application
   * and use that configuration.
   *
   * This can return either a Child or Main configuration.
   */
  static infer({
    directory,
    filePath,
    meta,
    cookies,
    options
  } = {}) {
    if (filePath && meta) {
      return MicrofrontendsServer.fromFile({
        filePath,
        cookies,
        meta,
        options
      });
    }
    try {
      const packageRoot = findPackageRoot(directory);
      const { name: appName } = getApplicationContext({ packageRoot });
      const configMeta = meta ?? { fromApp: appName };
      const maybeConfig = findConfig({ dir: packageRoot });
      if (maybeConfig) {
        return MicrofrontendsServer.fromFile({
          filePath: maybeConfig,
          cookies,
          meta: configMeta,
          options
        });
      }
      const repositoryRoot = findRepositoryRoot();
      const isMonorepo2 = isMonorepo({ repositoryRoot });
      if (isMonorepo2) {
        const defaultPackage = findDefaultMicrofrontendsPackage({
          repositoryRoot,
          applicationName: appName
        });
        const maybeConfigFromDefault = findConfig({ dir: defaultPackage });
        if (maybeConfigFromDefault) {
          return MicrofrontendsServer.fromFile({
            filePath: maybeConfigFromDefault,
            cookies,
            meta: configMeta,
            options
          });
        }
      }
      throw new Error("Unable to infer");
    } catch (e) {
      throw new MicrofrontendError(
        "Unable to locate and parse microfrontends configuration",
        { cause: e, type: "config", subtype: "inference_failed" }
      );
    }
  }
  /*
   * Generates a MicrofrontendsServer instance from a file.
   */
  static fromFile({
    filePath,
    cookies,
    meta,
    options
  }) {
    try {
      const configJson = fs6.readFileSync(filePath, "utf-8");
      const config = MicrofrontendsServer.validate(configJson);
      if (!isMainConfig(config) && options?.resolveMainConfig) {
        const repositoryRoot = findRepositoryRoot();
        const isMonorepo2 = isMonorepo({ repositoryRoot });
        if (isMonorepo2) {
          const packagePath = findPackagePath({
            repositoryRoot,
            name: config.partOf
          });
          if (!packagePath) {
            throw new MicrofrontendError(
              `Could not find default application "${config.partOf}" in the repository`,
              { type: "config", subtype: "not_found" }
            );
          }
          const maybeConfig = findConfig({ dir: packagePath });
          if (!maybeConfig) {
            throw new MicrofrontendError(
              `Could not find microfrontends configuration in ${packagePath}`,
              { type: "config", subtype: "not_found" }
            );
          }
          return MicrofrontendsServer.fromMainConfigFile({
            filePath: maybeConfig,
            overrides: cookies ? parseOverrides(cookies) : void 0
          });
        }
      }
      return new MicrofrontendsServer({
        config,
        overrides: cookies ? parseOverrides(cookies) : void 0,
        meta
      });
    } catch (e) {
      throw MicrofrontendError.handle(e, {
        fileName: filePath
      });
    }
  }
  /*
   * Generates a MicrofrontendMainConfig instance from a file.
   */
  static fromMainConfigFile({
    filePath,
    overrides
  }) {
    try {
      const config = fs6.readFileSync(filePath, "utf-8");
      const validatedConfig = MicrofrontendsServer.validate(config);
      if (!isMainConfig(validatedConfig)) {
        throw new MicrofrontendError(
          `${filePath} is not a main microfrontend config`,
          {
            type: "config",
            subtype: "invalid_main_path"
          }
        );
      }
      const [defaultApplication] = Object.entries(validatedConfig.applications).filter(([, app]) => isDefaultApp(app)).map(([name]) => name);
      if (!defaultApplication) {
        throw new MicrofrontendError(
          "No default application found. At least one application needs to be the default by omitting routing.",
          { type: "config", subtype: "no_default_application" }
        );
      }
      return new MicrofrontendsServer({
        config: validatedConfig,
        overrides,
        meta: { fromApp: defaultApplication }
      });
    } catch (e) {
      throw MicrofrontendError.handle(e, {
        fileName: filePath
      });
    }
  }
};

// src/vite/detect-framework.ts
import { existsSync } from "node:fs";
import { join as join2 } from "node:path";
import { cwd } from "node:process";
function detectFramework() {
  if (existsSync(join2(cwd(), "svelte.config.js"))) {
    return "sveltekit";
  }
  if (existsSync(join2(cwd(), "react-router.config.js")) || existsSync(join2(cwd(), "react-router.config.ts"))) {
    return "react-router";
  }
  return "unknown";
}

// src/vite/index.ts
function microfrontends(opts) {
  const { name: fromApp } = getApplicationContext();
  const microfrontendsObj = MicrofrontendsServer.infer({
    meta: {
      fromApp
    }
  });
  const app = microfrontendsObj.config.getApplication(fromApp);
  if (app.isDefault() && opts?.basePath) {
    throw new Error(
      "`basePath` can not be set for the default microfrontends application."
    );
  }
  if (opts?.basePath && !opts.basePath.startsWith("/")) {
    throw new Error("`basePath` must start with a `/`");
  }
  const additionalConfigOptions = {};
  const framework = detectFramework();
  if (!app.isDefault()) {
    if (opts?.basePath) {
      if (framework !== "react-router" || !process.env.VERCEL_ENV) {
        let basePath = opts.basePath;
        if (process.env.NODE_ENV === "production" && !basePath.endsWith("/")) {
          basePath = `${basePath}/`;
        }
        additionalConfigOptions.base = basePath;
      }
    } else if (framework !== "sveltekit") {
      additionalConfigOptions.experimental = {
        renderBuiltUrl(filename, { type }) {
          if (type === "asset") {
            return `/${app.getAssetPrefix()}/${filename}`;
          }
        }
      };
    }
    if (framework === "react-router") {
      additionalConfigOptions.build = {
        assetsDir: `./${opts?.basePath ?? app.getAssetPrefix()}`
      };
    }
  }
  if (app.development.local.port) {
    additionalConfigOptions.server = {
      port: app.development.local.port
    };
  }
  return {
    name: "vite-plugin-vercel-microfrontends",
    config: () => {
      return {
        ...additionalConfigOptions,
        define: {
          "import.meta.env.MFE_CURRENT_APPLICATION": JSON.stringify(app.name),
          "import.meta.env.MFE_CONFIG": JSON.stringify(
            microfrontendsObj.config.getConfig()
          )
        }
      };
    }
  };
}
export {
  microfrontends
};
